### ISSUE TITLE
Using "oneOf:" in OpenAPI Spec for /sc/ask Causes Problems in Java

### ISSUE THREAD

**Issue by Sophietje** [TNO/knowledge-engine#347]
In GitLab by @jarne.kerkaert.inetum-realdolmen.world on Jan 7, 2022, 10:59

In our services we generate the client classes for communicating with the KE using the [OpenAPI Generator Tool](https://github.com/OpenAPITools/openapi-generator). One change in the spec document causes a problem when trying to generate Java code:

```
requestBody:
        required: true
        content:
          application/json; charset=UTF-8:
            schema:
              oneOf:
                - $ref: '#/components/schemas/RecipientAndBindingSet'
                - $ref: '#/components/schemas/BindingSet'
```

The above code comes from the '/sc/ask' endpoint, it defines the body of the request to be either one of two possible structures (`RecipientAndBindingSet` or `BindingSet`). This particular thing might be possible to define in certain non-statically typed languages like Python or Javascript, but it causes a problem in Java.

This is an example of what the tool generates in Java (Notice the **UNKNOWN_BASE_TYPE**, which is invalid):

```
public AskResult scAskPost(String knowledgeBaseId, String knowledgeInteractionId, UNKNOWN_BASE_TYPE UNKNOWN_BASE_TYPE) throws RestClientException {
        return scAskPostWithHttpInfo(knowledgeBaseId, knowledgeInteractionId, UNKNOWN_BASE_TYPE).getBody();
    }
```

I don't know what the best solution to this would be, perhaps creating two endpoints would work. For now I've simply removed the `RecipientAndBindingSet` part, as we don't use it.

**Comment by Sophietje** [comment]
In GitLab by @barry.nouwt.tno.nl on Jan 25, 2022, 08:08

Hi @jarne.kerkaert.inetum-realdolmen.world, 

thanks for reporting this. The usage of `oneOf` Open API keyword here indeed seems to be incompatible with generating client Java code. In general, generating Java client code based on an Open API specification gave rise to several issues: #226, #127. I am a bit unsure about what to do with these issues, since there isn't always an easy solution. In the above case, for example, we introduced this `oneOf` keyword because we wanted to remain backwards compatible when we introduced the Recipient Selector.

Your solution of not including the RecipientAndBindingSet part is not ideal, but I am not sure how to solve it otherwise without breaking backward compatibility. Do you have any ideas?

Best regards,

Barry

**Comment by Sophietje** [comment]
In GitLab by @jarne.kerkaert.inetum-realdolmen.world on Jan 28, 2022, 08:42

Hey @barry.nouwt.tno.nl ,

I think that splitting the endpoint into two might be a solution (only in the openapi spec, not in the KE code). It should be possible to create two endpoints with different names, different bodies, but pointing to the same endpoint.

People who have already implemented a solution using a generated HTTP client with Python or Javascript will get some errors though. That is the only problem I currently see.

I'm currently quite swamped, but I could try this out if I do find some time.

**Comment by Sophietje** [comment]
In GitLab by @jarne.kerkaert.inetum-realdolmen.world on Feb 2, 2022, 12:19

@barry.nouwt.tno.nl I notice now that what I'm suggesting is impossible, apparently the entries in the spec are identified by URI, so they can not be duplicated.

**Comment by Sophietje** [comment]
In GitLab by @michel.hours.inetum.world on Feb 3, 2022, 08:29

Hi @barry.nouwt.tno.nl, is it possible for you to implement a workaround by adding a new supplementary endpoint with a new URI in your OpenAPI specification ?

**Comment by Sophietje** [comment]
In GitLab by @jarne.kerkaert.inetum-realdolmen.world on Feb 3, 2022, 09:41

@barry.nouwt.tno.nl 

I found a possible solution in this thread: https://github.com/OpenAPITools/openapi-generator/issues/15. Though it involves creating a parent model that both the `RecipientAndBindingSet` and `BindingSet` inherit from:

```
BindingSet:
      type: array
      items:
        nullable: false
        type: object
        properties:
          item:
            type: string
    RecipientAndBindingSet:
      type: object
      required: [recipientSelector, bindingSet]
      properties:
        recipientSelector:
          $ref: '#/components/schemas/RecipientSelector'
        bindingSet:
          $ref: '#/components/schemas/BindingSet'
    RecipientSelector:
      type: object
      properties:
        "singleKnowledgeBase":
          type: string
    RecipientSelectorAndBindingSetParent:
      type: object
      properties:
        type:
          type: string
          enum: [ 'recipientAndBindingSet', 'bindingSet' ]
      discriminator:
        propertyName: type
        mapping:
          recipientAndBindingSet: '#/components/schemas/RecipientAndBindingSet'
          bindingSet: '#/components/schemas/BindingSet'
```

This solution is quite complex though and also requires that the knowledge engine implements this kind of structure. This solution also breaks backwards compatibility, as the type of the body is different. I'm not sure how important it is to be backwards compatible in this case, but to me it looks best that the endpoint is split into two.

I think that in any case, anyone who wants to implement the recipient selector will have to change their code.

I really hope that this gets fixed somehow, as this is currently a blocking issue for us. As long as we can't use this feature, we **can not start collecting data** from our partners in the French pilot.

**Comment by Sophietje** [comment]
In GitLab by @barry.nouwt.tno.nl on Feb 4, 2022, 13:40

Hi everyone, this morning we put some effort into this issue, but unfortunately we did not really find a satisfying solution.

We were able to reproduce the `UNKNOWN_BASE_TYPE` behaviour above in our own environment, but adding a `RecipientSelectorAndBindingSet*Parent*` did not solve the issue for us. We also considered making the `BindingSet` an `object` instead of an `array`. This would allow us to merge the `BindingSet` and `RecipientAndBindingSet` into a single one, but breaks backward compatibility too much. Also adding an additional path (for example `POST /sc/ask_with_recipient`) seems a bit overkill and complicates the REST API (see also: https://gitlab.inesctec.pt/interconnect/generic-adapter/-/issues/67). So, the conclusion is that we unfortunately do not seem to be able to solve this within the REST API specification.

Would any of solutions below be an option for you?
- instead of using the OpenAPI Generator, implement a HTTP client for REST Developer API of the Knowledge Engine yourself. We also started with generating part of the server-side code using OpenAPI generator, but we have mostly moved away from it because it was not flexible enough. Although OpenAPI Generator seems handy at first, we think it is too inconsistent to be really useful.
- if you are accessing the Knowledge Engine from Java, it might be worth the investment to use the Java Developer API for the Knowledge Engine instead of the REST Developer API. This API is more flexible and probably also faster. It does not work with the Generic Adapter, but I am unsure whether you use that component (since it has its own REST API specification).
- generate the client code once using the OpenAPI Generator and manually fix the UNKNOWN_TYPE bug by replacing its occurrances with type Object. This worked in our demo environment, but it does mean future changes to the REST API specification need to be manually added. We also had to overwrite the default 'Accept: application/json; charset=UTF-8' to 'Accept: */*' in `SmartConnectorLifeCycleApi#317` to make it work.

Does this help you? Or should we plan a session to try and find a solution together.

Best regards,

Barry
